gadget/update.go

```diff
@@ -132,3 +198,97 @@ func canUpdateVolume(from *PositionedVolume, to *PositionedVolume) error {
 	}
 	return nil
 }
+
+type updatePair struct {
+	from *PositionedStructure
+	to   *PositionedStructure
+}
+
+func resolveUpdate(oldVol *PositionedVolume, newVol *PositionedVolume) (updates []updatePair) {
+	for j, oldStruct := range oldVol.PositionedStructure {
+		newStruct := newVol.PositionedStructure[j]
+		// update only when new edition is higher than the old one; boot
+		// assets are assumed to be backwards compatible, once deployed
+		// are not rolled back or replaced unless a higher edition is
+		// available
+		if newStruct.Update.Edition > oldStruct.Update.Edition {
+			updates = append(updates, updatePair{
+				from: &oldVol.PositionedStructure[j],
+				to:   &newVol.PositionedStructure[j],
+			})
+		}
+	}
+	return updates
+}
+
+type Updater interface {
+	// Update applies the update or errors out on failures
+	Update() error
+	// Backup prepares a backup copy of data that will be modified by
+	// Update()
+	Backup() error
+	// Rollback restores data modified by update
+	Rollback() error
+}
+
+type nopUpdater struct{}
+
+func (n *nopUpdater) Update() error {
+	return nil
+}
+
+func (n *nopUpdater) Backup() error {
+	return nil
+}
+
+func (n *nopUpdater) Rollback() error {
+	return nil
+}
+
+// TODO: replace placeholder implementation with real one
+var updaterForStructure = func(_ *PositionedStructure, rootDir, rollbackDir string) (Updater, error) { return &nopUpdater{}, nil }
+
+func applyUpdates(new GadgetData, updates []updatePair, rollbackDir string) error {
+	updaters := make([]Updater, 0, len(updates))
+
+	for _, one := range updates {
+		up, err := updaterForStructure(one.to, new.RootDir, rollbackDir)
+		if err != nil {
+			return fmt.Errorf("cannot prepare updater: %v", err)
+		}
+		updaters = append(updaters, up)
+	}
+
+	for _, one := range updaters {
+		if err := one.Backup(); err != nil {
+			return fmt.Errorf("cannot backup structure: %v", err)
```

Comment:
I wonder if in external message we should always have "volume structure" and not just structure

Reply:
Added info about the actual structure that failed too.